Frameworks/Foundation/NSOperationQueue.mm

```diff
@@ -38,27 +38,52 @@ static inline long _QOSClassForNSQualityOfService(NSQualityOfService quality) {
         case NSQualityOfServiceUtility:
             return QOS_CLASS_UTILITY;
         case NSQualityOfServiceBackground:
-            // currently generates a warning:
-            // 'QOS_CLASS_BACKGROUND' is deprecated: QOS_CLASS_BACKGROUND is the same as DISPATCH_QUEUE_PRIORITY_LOW on WinObjC
-            // should go away once libdispatch is updated
+// currently generates a warning:
+// 'QOS_CLASS_BACKGROUND' is deprecated: QOS_CLASS_BACKGROUND is the same as DISPATCH_QUEUE_PRIORITY_LOW on WinObjC
+// TODO #2241: should go away once libdispatch is updated
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wdeprecated-declarations"
             return QOS_CLASS_BACKGROUND;
+#pragma clang diagnostic pop
         case NSQualityOfServiceDefault:
         default:
             return QOS_CLASS_DEFAULT;
     }
 }
 
-// Used for [NSOperationQueue currentQueue] - when an NSOperation is executed through an NSOperationQueue,
-// the current queue is stored in the thread dictionary using this key.
-static const NSString* const _NSOperationQueueCurrentQueueKey = @"_NSOperationQueueCurrentQueueKey";
+// Private helper that compares two NSOperations and orders the higher-priority one first
+static NSInteger _compareOperationPriority(NSOperation* op1, NSOperation* op2, void* context) {
+    NSOperationQueuePriority pri1 = op1.queuePriority;
+    NSOperationQueuePriority pri2 = op2.queuePriority;
+    if (pri1 > pri2) {
+        return NSOrderedAscending;
+    } else if (pri1 < pri2) {
+        return NSOrderedDescending;
+    }
+    return NSOrderedSame;
+}
+
+// STL compare for NSOperation*, sorting by queuePriority, putting the greatest priority first
+struct _NSOperation_ComparePrioritySTL {
+    bool operator()(NSOperation* lhs, NSOperation* rhs) {
+        return lhs.queuePriority > rhs.queuePriority;
+    }
+};
+
+// Used for [NSOperationQueue currentQueue]
+thread_local static StrongId<NSOperationQueue*> tls_currentQueue;
+
+// Used for key-value observing on held operations - faster than comparing keypaths
+static void* _NSOperationQueue_IsFinishedContext = &_NSOperationQueue_IsFinishedContext;
```

Comment:
nit: you can use use `static char _xyz` and then `&_xyz` later; no need for initialization (or emitting a binary relocation that requires updating a pointer :smile:)